### DISCUSSION TITLE
How to use renderA11yString

### DISCUSSION THREAD

**Discussion by mustafasalahuldin** [KaTeX/KaTeX#3819] [Q&A] [answered]
Hello
I came across this extension 'renderA11yString' in the github folder. I'm too much interested to use it in Typescript application. It has a lot of issues specially when I try to import it. Can someone advise please?

**Comment by edemaine** [comment]
I believe either of the following imports should work:

```js
// ESM
import renderA11yString from "katex/dist/contrib/render-a11y-string.mjs"
// CJS
renderA11yString = require("katex/dist/contrib/render-a11y-string.js")
```

Can you try and report back?

**Reply by mustafasalahuldin** [reply]
Now I can use the module, but still getting this error:
`katex.js:18555  Uncaught TypeError: KaTeX can only parse string typed expression
    at Xr (katex.js:18555:11)
    at Object.__parse (katex.js:18642:10)
    at y.default (render-a11y-string.js:869:59)
    at a.onend (app.ts:1074:40)
Xr @ katex.js:18555
__parse @ katex.js:18642
y.default @ render-a11y-string.js:869
a.onend @ app.ts:1074`

When I'm trying to parse something like this:
"Here's an example of an inline complex math equation: $(2+3i) \cdot (4-2i) = 14 + 10i$"

**Reply by edemaine** [reply]
I'm not sure how you're using it, but it looks like you're not passing a string into the function.  Here's an example successful usage:

```node
> renderA11yString = require("katex/dist/contrib/render-a11y-string.js")
[Function: renderA11yString]
> renderA11yString("(2+3i) \\cdot (4-2i) = 14 + 10i")
'left parenthesis, 2, plus, 3, i, right parenthesis, dot, left parenthesis, 4, minus, 2, i, right parenthesis, equals, 14, plus, 10, i'
```

**Reply by mustafasalahuldin** [reply]
@edemaine It worked because you used the expression only without the text. Please use the whole sentence as I quoted it. Usually when I use the render function and there is $ in the expression, it fails showing you can't use $ in math mode.
I ca't predict the delimiter, so either I need a way to configure it, or Katex should be smart enough to detect it.😄

**Reply by edemaine** [reply]
renderA11yString is designed for a single math expression. (I'm not even sure what the output should be for text...) You'll have to write your own math expression extractor, or use `splitAtDelimiters` from the auto-render extension.

**Reply by mustafasalahuldin** [reply]
@edemaine  So you mean, splitAtDelimiters will return array of expressions, then I pass them to renderA11yString?

**Reply by mustafasalahuldin** [reply, marked as answer]
I discovered that renderA11yString is generating `<annotation encoding="application/x-tex">` element, which was useful in my case.
I used the document.getElementByTagName('annotation ') to find the readings.

**Reply by vincerubinetti** [reply]
I've found that the `renderA11yString` has too many "not implemented yet" errors to be used effectively in a production app.

I also usually haven't seen the contents of `<annotation>` to be something readable, i.e. the output of the `renderA11yString` function. I usually just see the original latex there. But perhaps that's what it falls back to when the function errors, and all of my latex just happens to not work with the function.

Anyway, to get some kind of SR announcement working, and to get Axe checks on my site to pass, I've had to hoist up the contents of `<annotation>` to the root katex element:

```js
document.querySelectorAll("annotation").forEach((element) => {
  const content = element.textContent ?? "";
  const root = element.closest(".katex");
  root?.setAttribute("aria-label", content);
  root?.setAttribute("role", "img");
});
```
